src/suricata-plugin.h

```diff
@@ -41,16 +41,30 @@ typedef struct SCPlugin_ {
     void (*Init)(void);
 } SCPlugin;
 
+typedef struct SCPluginFileOptions_ {
+    bool threaded;
+    const char *filename; /* Configured file name */
+    const char *log_dir;
+    const char *append;
+} SCPluginFileOptions;
+
 /**
  * Structure used to define a file type plugin.
  *
  * Currently only used by the Eve output type.
  */
 typedef struct SCPluginFileType_ {
     char *name;
-    int (*Open)(ConfNode *conf, void **data);
-    int (*Write)(const char *buffer, int buffer_len, void *ctx);
+    /* Called on first open; no other calls will be made if unthreaded */
+    int (*Open)(ConfNode *conf, SCPluginFileOptions *opt, void **ctx);
+    /* Write - Called on each write to the object */
+    int (*Write)(const char *buffer, int buffer_len, void *ctx, void *thread_ctx);
+    /* Close - Called on final close */
     void (*Close)(void *ctx);
+    /* Open - Called for each thread using the file object */
+    int (*ThreadOpenFunc)(void *ctx, int thread_id, void **thread_ctx);
+    /* Close - Called for each thread using the file object */
+    int (*ThreadCloseFunc)(void *ctx, void *thread_ctx);
```

Comment:
Naming... In Suricata was use the Func function for some funcs, for function pointer variable names and for function point types.  I propose that in the plugin API we save the "Func" suffext for function type declartions, but not for the variable names..

We could `typedef int ThreadOpenFunc(...)`, but for end-users its often just easier to see the function definitions right in the initialization structure as well.

Just thinking out loud here.. Thoughts?

Reply:
I don't mind changing the name to avoid confusion .... `ThreadOpen` make sense?